### ISSUE TITLE
HttpClient concerts HEAD requests to GET request

### ISSUE THREAD

**Issue by bzd3y** [dotnet/runtime#109791]
### Android framework version

net8.0-android

### Affected platform version

SDK 8.0.400, VS 17.11.35327.3

### Description

On Android, the default `HttpMessageHandler` (`HttpClientHandler`?) does not correctly follow redirects with a HEAD request and converts the request to a GET.

The following code can be used to reproduce this on Android. Using the same code on Windows results in different/correct behavior.

```csharp
private async Task MakeRequestAsync(string url, HttpMessageHandler handler = null)
{
	using HttpRequestMessage request = new(HttpMethod.Head, url);
	HttpClient client = handler switch
	{
		null => new(),
		_ => new(handler)
	};
	using HttpResponseMessage response = await client.SendAsync(request);

	Console.WriteLine($"--- Request ---");
	Console.WriteLine(response.RequestMessage);
	Console.WriteLine($"--- Response ---");
	Console.WriteLine(response);
}

private async Task MakeRequestsAsync()
{
	Console.WriteLine("--- using default HttpMessageHandler ---");

	await MakeRequestAsync("https://github.com/dotnet/android/raw/refs/heads/main/.editorconfig"); // redirects, gets converted to GET
	await MakeRequestAsync("https://raw.githubusercontent.com/dotnet/android/refs/heads/main/.editorconfig"); // target of redirect, stays HEAD

	Console.WriteLine("--- using SocketsHttpHandler ---");

	await MakeRequestAsync("https://github.com/dotnet/android/raw/refs/heads/main/.editorconfig", new SocketsHttpHandler()); // redirects, stays HEAD
	await MakeRequestAsync("https://raw.githubusercontent.com/dotnet/android/refs/heads/main/.editorconfig", new SocketsHttpHandler()); // target of redirect, stays HEAD
}
```

On Android, if a `SocketsHttpHandler` is used as the handler then it seems to follow the redirects correctly and perform a HEAD request to the location indicated in the initial redirect response. I don't know if there is a reason to avoid using `SocketsHttpHandler`, but the IDE does give a warning for [CsWinRT1030](https://github.com/microsoft/CsWinRT/blob/master/src/Authoring/WinRT.SourceGenerator/AnalyzerReleases.Shipped.md) in Android/MAUI.

If the default handler is used to follow redirects manually then it seems like it correctly performs a HEAD request initially and returns a `Location` header that can be followed with another HEAD request. ~~I think this is going to have to be my workaround for now, although the fact that this can't be set per request might make that difficult.~~ Never mind, I'm going to have to go with `SocketsHttoHandler`. The documentation indicates that that is supposed to be the default anyway, so I'm not sure what is going on there. I also don't seem to get the warning about CsWinRT1030 in my actual project or any warnings. I do not have `<AllowUnsafeBlocks>True</AllowUnsafeBlocks>` in my project file, but maybe it depends on how AOT is configured. So I'm going with this.

I also notice the same behavior as https://github.com/dotnet/android/issues/8848 and https://github.com/dotnet/android/issues/4695 and would not be surprised if these issues are related, where even with `HttpCompletionOption.ResponseHeadersRead` set on the request, the `HttpClient`/`HttpMessageHandler` may be doing a full request. The reason I suspect that is happening is that if the request is done on the main thread, the `SendAsync()` will crash, per dotnet/android#4695, even if I never read the `Contents` or even reference it, unless `ConfigureAwait(false)` is used on the `SendAsync()` call.

The bottom line is that both of the ways to tell `HttpClient` to "just get the headers" do not work on Android.

### Steps to Reproduce

The above code can be added to a new MAUI project or a Windows Console.

### Did you find any workaround?

Yes, separate requests or using `SocketsHttpHandler`, though the latter apparently has some caveats.

### Relevant log output

```shell

```

**Comment by julealgon** [comment]
> ...does not correctly follow redirects with a HEAD request

What redirect HTTP Status code are you returning from the server?

Could this behavior be related to [the difference between 302 and 307](https://stackoverflow.com/questions/2068418/whats-the-difference-between-a-302-and-a-307-redirect)?

**Comment by bzd3y** [comment]
@julealgon I'm not returning anything from the server. I'm making requests to a server out of my control. That server (GitHub, in this case) is returning 302 according to DotNet and cURL (and cURL follows the redirect correctly as well).

I guess it is possible this could somehow be related, but I don't see how and I'm pretty confident it would be a bug either way (even if it is just that an old handler is being used as the default when it shouldn't be).

**Comment by julealgon** [comment]
@bzd3y to be clear I was not implying this wasn't a bug. Was just curious if it was related to the differences between those 2 particular status codes since I've read about them recently by coincidence due to another issue we had on our side.

**Comment by bzd3y** [comment]
@julealgon I know, no worries. I was just differentiating between the two "paths", though maybe not clearly. One being fixing the handler and the other being fixing which handler is being used.
